### ISSUE TITLE
How to read / write data ? 

### ISSUE THREAD

**Issue by nukadelic** [Unity-Technologies/barracuda-release#12]
Got the following model:
![image](https://user-images.githubusercontent.com/6582633/71777622-0b290a80-2fd5-11ea-9540-65aeae3ec72e.png)

Logging the inputs & outputs shapes in console can see 
![image](https://user-images.githubusercontent.com/6582633/71777646-5e9b5880-2fd5-11ea-9851-f429b53960a8.png)

But the default log info is a bit different ( there is no negative size )
![image](https://user-images.githubusercontent.com/6582633/71777656-796dcd00-2fd5-11ea-9170-59e254d1ae48.png)

I used this sample code to create an auto encoder in Keras:
https://github.com/jg-fisher/autoencoder/blob/master/ffae.py

Here is my attempt so far, all i can see are unchanging zeros values as the output :
````cs
var inputs = new Dictionary<string, Tensor>();

float[][] jagged_input_data = new float[][]
{
    new float[] { 1f },
    new float[] { 0f },
    new float[] { 1f }
};

var input_1_1 = new Tensor(-1, 1, 1, 3, jagged_input_data);
inputs["input_1_1"] = input_1_1;

worker.Execute(inputs);
        
Tensor output = worker.Fetch("dense_1_1/Relu");
float[] data = new float[1 * 1 * 3];

output.data.Upload(data, 0, 3 );

Debug.Log(string.Join(",", data) );
````
Am i doing something wrong with the dimensions ?

**Comment by nukadelic** [comment]
I finally managed to get some output via:
```cs
output = worker.Fetch(model.outputs[0]);
float[] data = output.data.Download( 2 );
```
But the model accuracy seems to be off
![image](https://user-images.githubusercontent.com/6582633/71816761-38011e80-30b6-11ea-8b0a-6c5348e36736.png)
Here is my test script:
```cs
public class barracuda_test_encoder : MonoBehaviour
{
    static float[] TEST_INPUTS = 
    {
        1,1,1,
        1,1,2,
        ...
    };

    static float[] TEST_OUTPUTS =
    { 
        0.5530f, 2.4361f,
        0.0000f, 2.4961f,
        ...
    };

    //public NNModel modelSource;

    Model model;
    IWorker worker;
    Dictionary<string, Tensor> inputs;
    Tensor output;

    void Start()
    {
        bool debug_log = false;
        model = ModelLoader.LoadFromStreamingAssets("encoder_model.nn");
        worker = BarracudaWorkerFactory.CreateWorker(BarracudaWorkerFactory.Type.ComputePrecompiled, model, debug_log);
        StartCoroutine(TestWorker());
    }

    IEnumerator TestWorker()
    {
        for( var i = 0; i < TEST_INPUTS.Length / 3; i ++ )
        {
            string line = "";

            yield return new WaitForSeconds(0.1f);

            var idx_input = i * 3;
            var idx_output = i * 2;

            inputs = new Dictionary<string, Tensor>();
            
            float[][] jagged_input_data = new float[][]
            {
                new float[] { TEST_INPUTS[ idx_input + 0 ] },
                new float[] { TEST_INPUTS[ idx_input + 1 ] },
                new float[] { TEST_INPUTS[ idx_input + 2 ] }
            };

            line += "Input data = ( ";

            foreach (var data_point in jagged_input_data)

                line += string.Format("{0,4:##0.00}, ", data_point[0]);

            line += " ) ";

            var input_1_1 = new Tensor(1, 3, jagged_input_data );

            inputs.Add("input_1_1", input_1_1);

            var coroutine = worker.ExecuteAsync(inputs);
            yield return StartCoroutine(coroutine);

            output = worker.Fetch(model.outputs[0]);
            float[] data = output.data.Download( 2 );

            line += " --> Decoder output = ( ";
            foreach (var data_point in data) line += string.Format("{0,4:##0.00}, ", data_point);
            line += " ) ";
            line += " [vs] Keras output = ( ";
            line += string.Format("{0,4:##0.00}", TEST_OUTPUTS[ idx_output + 0 ] ) + ", ";
            line += string.Format("{0,4:##0.00}", TEST_OUTPUTS[ idx_output + 1 ] );
            line += " )";

            Debug.Log(line);

            yield return new WaitForSeconds(0.01f);

            foreach (var input in inputs) input.Value.Dispose();
            output.Dispose();
        }

        worker.Dispose();
    }
}
```

**Comment by nukadelic** [comment]
fixed
```cs
float[] input_data = new float[] {
    TEST_INPUTS[ idx_input + 0 ],
    TEST_INPUTS[ idx_input + 1 ],
    TEST_INPUTS[ idx_input + 2 ]
};
```

**Comment by ghost** [comment]
Hello @nukadelic !

Have you achieved to put your own model into an agent in Unity?

I tried to just dragging my own model here (.nn model converted from .onnx):

![Captura](https://user-images.githubusercontent.com/44144188/74102679-15937280-4b46-11ea-8935-ff69d5f6ddb6.PNG)

But I got this following error:

"NullReferenceException: Object reference not set to an instance of an object"

![Captura2](https://user-images.githubusercontent.com/44144188/74102698-4a072e80-4b46-11ea-8416-e347e6d2cd7f.png)

Any suggeriment of what I am doing wrong?

Thank you :)

**Comment by nukadelic** [comment]
yes i did, @noocin i used barracuda 0.3.2 and made this colab for the steps i took to make it happened: https://colab.research.google.com/drive/1xqxBPzeJughcBURep810yLW685IfbA3T

**Comment by ghost** [comment]
Thank you very much @nukadelic, it seems very complicated hahaha

I just followed these steps to convert a model from Pytorch (.pth) to Barracuda (.nn):

Convert to pytorch to onnx
https://pytorch.org/docs/stable/onnx.html then
to nn with onnx_to_barracuda
https://github.com/mantasp/barracuda-release/tree/release/0.2.4/Tools

I think the problem is how I name the inputs and outputs...

But I think I will wait for barracuda support on that.
